client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -1282,18 +1302,58 @@ private StreamSegments getStreamSegments(final SegmentRanges ranges, long reques
+        Segment segment = Segment.fromScopedName(qualifiedSegmentName);
+        long segmentId = segment.getSegmentId();
+        CachedPravegaNodeUri nodeUri = getSegmentEndpointFromCache(segmentId);
+        //Read from cache if the segment endpoint already exists and the refresh interval is not expired.
+        if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() <= CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            log.info("Fetching the endpoint details of segment {} from cache", segmentId);
+            return nodeUri.getPravegaNodeUri();
+        } else if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() > CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            CompletableFuture<PravegaNodeUri> existingNodeInfo = nodeUri.getPravegaNodeUri();
+            // Trigger a background call and refresh the cache.
+            endPointCacheMap.put(segment.getSegmentId(), new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(qualifiedSegmentName)));
+            return existingNodeInfo;
+        } else {
+            endPointCacheMap.put(segment.getSegmentId(), new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(qualifiedSegmentName)));
+            return endPointCacheMap.get(segment.getSegmentId()).getPravegaNodeUri();
+        }
+    }
+
+    public void updateStaleValueInCache(String segmentName, PravegaNodeUri errNodeUri) {
+        Exceptions.checkNotNullOrEmpty(segmentName, "segmentName");
+        Segment segment = Segment.fromScopedName(segmentName);
+        long segmentId = segment.getSegmentId();
         final long requestId = requestIdGenerator.get();
-        long traceId = LoggerHelpers.traceEnter(log, "getEndpointForSegment", qualifiedSegmentName, requestId);
+        long traceId = LoggerHelpers.traceEnter(log, "updateStaleValueInCache", segmentName, errNodeUri, requestId);
+        CachedPravegaNodeUri cachedNode = getSegmentEndpointFromCache(segmentId);
+        cachedNode.getPravegaNodeUri().thenAccept(cachedNodeUri -> {
+            if (cachedNodeUri.getEndpoint().equals(errNodeUri.getEndpoint()) && cachedNodeUri.getPort() == errNodeUri.getPort()) {
+                // enforce cache refresh in case of stale value
+                log.debug(requestId, "Refreshing stale value in cache for segment {}", segmentId);
+                endPointCacheMap.put(segment.getSegmentId(), new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(segmentName)));
+            }
+        }).whenComplete((x, e) -> {
+            if (e != null) {
+                log.warn("updateStaleValueInCache failed for segment {}: ", segmentName, e);
+                endPointCacheMap.remove(segmentId);
+            }
+            LoggerHelpers.traceLeave(log, "updateStaleValueInCache", traceId, requestId);
+        });
+    }
 
+    private CompletableFuture<PravegaNodeUri> getPravegaNodeUri(String qualifiedSegmentName) {
+        Segment segment = Segment.fromScopedName(qualifiedSegmentName);
```

Comment:
Are we using this **segment** in the code ?

Reply:
yes, same code block it is used.

Reply:
The caller in each case has a `Segment` object. So rather than converting it to a string and then parsing that string, we could just make `Segment segment` a parameter.

Reply:
Incorporated changes